src/engine/BMInterfaceTournament.php

```diff
@@ -0,0 +1,987 @@
+                 'AND tourn_id = :tourn_id2 '.
+                 'AND player_id = :player_id';
+
+        $statement = self::$conn->prepare($query);
+        $parameters = array(':tourn_id1' => $tournId,
+                            ':tourn_id2' => $tournId,
+                            ':player_id' => $userId,
+                            ':n_players' => $nPlayers);
+        return self::$db->update_and_report_if_changed($query, $parameters);
+    }
+
+    /**
+     * Attempt to cancel a tournament
+     *
+     * @param int $userId
+     * @param int $tournId
+     * @return bool
+     */
+    protected function cancel_tournament($userId, $tournId) {
+        try {
+            $tournament = $this->load_tournament($tournId);
+
+            if (!$this->validate_cancel_tournament($userId, $tournId, $tournament)) {
+                return NULL;
+            }
+
+            $success = $this->cancel_tournament_in_database($userId, $tournId);
+
+            if (!$success) {
+                // something's gone wrong between validation and attempting to remove the user
+                $tournament = $this->load_tournament($tournId);
+                $validation = $this->validate_cancel_tournament($userId, $tournId, $tournament);
+
+                if ($validation) {
+                    error_log('Paradoxical validation success in BMInterface::cancel_tournament');
+                    $this->set_message('Tournament cancel failed even though it should have succeeded');
+                }
+                return NULL;
+            }
+        } catch (BMExceptionDatabase $e) {
+            $this->set_message('Cannot cancel tournament because a player or tourn ID was not valid');
+            return NULL;
+        } catch (Exception $e) {
+            error_log(
+                'Caught exception in BMInterface::cancel_tournament: ' .
+                $e->getMessage()
+            );
+            $this->set_message("Tournament cancel failed: $e");
+            return NULL;
+        }
+
+        $this->update_tournament($tournId);
+        $this->set_message('You have successfully cancelled this tournament');
+        return TRUE;
+    }
+
+    /**
+     * Validate whether a player can cancel a tournament
+     *
+     * @param int $userId
+     * @param int $tournId
+     * @param BMTournament $tournament
+     * @return bool
+     */
+    protected function validate_cancel_tournament($userId, $tournId, $tournament) {
+        if (is_null($tournament)) {
+            $this->set_message("Tournament $tournId does not exist");
+            return NULL;
+        }
+
+        if ($tournament->tournamentState > BMTournamentState::JOIN_TOURNAMENT) {
+            $this->set_message('The tournament has already started.');
+            return NULL;
+        }
+
+        if ($tournament->creatorId !== $userId) {
+            $this->set_message('You did not create this tournament.');
+            return NULL;
+        }
+
+        return TRUE;
+    }
```

Comment:
We think it's okay to cancel a tournament that other people have joined, as long as it's still in "joining"?  That seems unfriendly, but maybe it's correct.  Do we have any sort of tournament UI for cancelled tournaments so people who joined can see what happened, or not yet?

(Flagging this design choice for reviewers.)

Reply:
I personally think that a tournament creator should always have the right to cancel a tournament that has not yet started.

Someone who has joined a tournament that is later cancelled can see this on their tournament overview page, under the heading Cancelled Tournaments.